src/ruby/lib/grpc/generic/active_call.rb

```diff
@@ -169,10 +169,13 @@ def receive_and_check_status
       batch_result = @call.run_batch(ops)
       unless @metadata_received
         @call.metadata = batch_result.metadata
-        @metadata_received = true
       end
       set_input_stream_done
       attach_status_results_and_complete_call(batch_result)
+    ensure
+      # Ensure we don't attempt to request the initial metadata again
+      # in case an exception occurs.
+      @metadata_received = true
```

Comment:
This becomes the opposite that of the original report that if `run_batch` throws before sent RECV_INITIAL_METADATA, we are setting metadata_received anyways.

Ideally we need to c-core to propagate the error from [completion queue](https://github.com/grpc/grpc/blob/v1.56.0/src/core/lib/surface/completion_queue.cc#L773) so here we know whether to metadata_received or not.

Given  that's a fairly large change in c-core, I incline to accept this to unblock the customer as I mentioned in the other comment in the issue, missing a RECV_INITIAL_METADATA doesn't really cause problems (for now at least).

@apolcyn Do you have any other concerns?